pkg/scd/operational_intents_handler.go

```diff
@@ -363,95 +363,132 @@ func (a *Server) UpdateOperationalIntentReference(ctx context.Context, req *rest
+		return nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "Invalid state for initial version: `%s`", params.State)
 	}
+	valid.ovn = ovn
 
-	subscriptionID := dssmodels.ID("")
 	if params.SubscriptionId != nil {
-		subscriptionID, err = dssmodels.IDFromOptionalString(string(*params.SubscriptionId))
+		valid.subscriptionID, err = dssmodels.IDFromOptionalString(string(*params.SubscriptionId))
 		if err != nil {
-			return nil, nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "Invalid ID format for Subscription ID: `%s`", *params.SubscriptionId)
+			return nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "Invalid ID format for Subscription ID: `%s`", *params.SubscriptionId)
 		}
 	}
 
 	// Check if a subscription is required for this request:
 	// OIRs in an accepted state do not need a subscription.
-	if state.RequiresSubscription() &&
-		subscriptionID.Empty() &&
+	if valid.state.RequiresSubscription() &&
+		valid.subscriptionID.Empty() &&
 		(params.NewSubscription == nil ||
 			params.NewSubscription.UssBaseUrl == "") {
-		return nil, nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "Provided Operational Intent Reference state `%s` requires either a subscription ID or information to create an implicit subscription", state)
+		return nil, stacktrace.NewErrorWithCode(dsserr.BadRequest, "Provided Operational Intent Reference state `%s` requires either a subscription ID or information to create an implicit subscription", valid.state)
+	}
+
+	return valid, nil
+}
+
+// upsertOperationalIntentReference inserts or updates an Operational Intent.
+// If the ovn argument is empty (""), it will attempt to create a new Operational Intent.
+func (a *Server) upsertOperationalIntentReference(ctx context.Context, authorizedManager *api.AuthorizationResult, entityid restapi.EntityID, ovn restapi.EntityOVN, params *restapi.PutOperationalIntentReferenceParameters,
+) (*restapi.ChangeOperationalIntentReferenceResponse, *restapi.AirspaceConflictResponse, error) {
+	// Note: validateAndReturnUpsertParams could be moved out of this method and only the valid params passed,
+	// but this requires some changes in the caller that go beyond the immediate scope of #1088 and can be done later.
+	upsertParams, err := a.validateAndReturnUpsertParams(authorizedManager, entityid, ovn, params)
+	if err != nil {
+		// Important note:
+		//   - stacktrace.Propagate sets NoCode as the code for the error
+		//   - subsequently, stacktrace.GetCode(err) only looks into the first error in the stack and returns NoCode if the error has no code
+		//   - therefore, we need to explicitly use stacktrace.PropagateWithCode if we want the calling logic
+		//   to be able to retrieve the code of the error
+		// TODO we can consider fixing this in the stacktrace library
+		return nil, nil, stacktrace.PropagateWithCode(err, stacktrace.GetCode(err), "Failed to validate Operational Intent Reference upsert parameters")
```

Comment:
Here I suggest making `validateAndReturnUpsertParams` purely a validation function by:
- move out of `validateAndReturnUpsertParams` the two cases of errors `dsserr.PermissionDenied`
- never returning any code within `validateAndReturnUpsertParams`
- if `err != nil` set the code `dsserr.BadRequest`

That makes it clear this function validates the input to create the op intent and any error implies bad input. That should also enable passing the manager instead of the whole `authorizedManager`.